core/src/error/mod.rs

```diff
@@ -234,6 +241,82 @@ impl Error {
     }
 }
 
+/// Child diagnostic methods
+#[cfg(feature = "diagnostics")]
+impl Error {
+    pub fn error<T: fmt::Display>(mut self, message: &T) -> Self {
```

Comment:
NIT: all this code is copy-pasted and could be macro-generated for the sake of maintainablity. By the way, [it is what they do in the `proc_macro` library](https://github.com/rust-lang/rust/blob/master/library/proc_macro/src/diagnostic.rs#L57). I don't know if it's worth the trouble though :)

Reply:
Fixed.

Reply:
I made this into a macro. The code is now much nicer to read - thanks for the suggestion.